types/src/v2/mod.rs

```diff
@@ -160,3 +176,126 @@ impl<'a> RpcParams<'a> {
 		self.parse::<[T; 1]>().map(|[res]| res)
 	}
 }
+
+/// [JSON-RPC parameters](https://www.jsonrpc.org/specification#parameter_structures)
+///
+/// If your type implement `Into<JsonValue>` call that favor of `serde_json::to:value` to
+/// construct the parameters. Because `serde_json::to_value` serializes the type which
+/// allocates whereas `Into<JsonValue>` doesn't in most cases.
+#[derive(Serialize, Debug)]
+#[serde(untagged)]
+pub enum JsonRpcParams<'a> {
+	/// No params.
+	NoParams,
+	/// Positional params (heap allocated)
+	Array(Vec<JsonValue>),
+	/// Positional params (slices)
+	ArrayRef(&'a [JsonValue]),
+	/// Params by name.
+	Map(BTreeMap<&'a str, JsonValue>),
+}
+
+impl<'a> From<BTreeMap<&'a str, JsonValue>> for JsonRpcParams<'a> {
+	fn from(map: BTreeMap<&'a str, JsonValue>) -> Self {
+		Self::Map(map)
+	}
+}
+
+impl<'a> From<Vec<JsonValue>> for JsonRpcParams<'a> {
+	fn from(arr: Vec<JsonValue>) -> Self {
+		Self::Array(arr)
+	}
+}
+
+impl<'a> From<&'a [JsonValue]> for JsonRpcParams<'a> {
+	fn from(slice: &'a [JsonValue]) -> Self {
+		Self::ArrayRef(slice)
+	}
+}
+
+/// Serializable [JSON-RPC object](https://www.jsonrpc.org/specification#request-object)
+#[derive(Serialize, Debug)]
+pub struct JsonRpcCallSer<'a> {
+	/// JSON-RPC version.
+	pub jsonrpc: TwoPointZero,
+	/// Name of the method to be invoked.
+	pub method: &'a str,
+	/// Request ID
+	pub id: u64,
+	/// Parameter values of the request.
+	pub params: JsonRpcParams<'a>,
+}
+
+impl<'a> JsonRpcCallSer<'a> {
+	/// Create a new serializable JSON-RPC request.
+	pub fn new(id: u64, method: &'a str, params: JsonRpcParams<'a>) -> Self {
+		Self { jsonrpc: TwoPointZero, id, method, params }
+	}
+}
+
+/// Serializable [JSON-RPC notification object](https://www.jsonrpc.org/specification#request-object)
+#[derive(Serialize, Debug)]
+pub struct JsonRpcNotificationSer<'a> {
+	/// JSON-RPC version.
+	pub jsonrpc: TwoPointZero,
+	/// Name of the method to be invoked.
+	pub method: &'a str,
+	/// Parameter values of the request.
+	pub params: JsonRpcParams<'a>,
+}
+
+impl<'a> JsonRpcNotificationSer<'a> {
+	/// Create a new serializable JSON-RPC request.
+	pub fn new(method: &'a str, params: JsonRpcParams<'a>) -> Self {
+		Self { jsonrpc: TwoPointZero, method, params }
+	}
+}
+
+/// JSON-RPC parameter values for subscriptions.
+#[derive(Deserialize, Debug)]
+pub struct JsonRpcNotificationParamsAlloc<T> {
+	/// Subscription ID
+	pub subscription: SubscriptionId,
+	/// Result.
+	pub result: T,
+}
+
+/// JSON-RPC notification response.
+#[derive(Deserialize, Debug)]
+pub struct JsonRpcNotifAlloc<T> {
+	/// JSON-RPC version.
+	pub jsonrpc: TwoPointZero,
+	/// Params.
+	pub params: JsonRpcNotificationParamsAlloc<T>,
+}
+
+/// Id of a subscription, communicated by the server.
+#[derive(Debug, PartialEq, Clone, Hash, Eq, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum SubscriptionId {
```

Comment:
I think we could kill this type and just use JsonValue as the subscription ID, alternatively reduce it to u64 as the servers does.